utils/lister.cc

```diff
@@ -61,23 +65,112 @@ future<> lister::scan_dir(fs::path dir, lister::dir_entry_types type, lister::sh
     });
 }
 
+static std::vector<directory_entry> list_directory(int fd) {
+    std::vector<directory_entry> ret;
+    static constexpr size_t buffer_size = 8192;
+    char buffer[buffer_size];
+
+    // From getdents(2):
+    struct linux_dirent64 {
+        ino64_t        d_ino;    /* 64-bit inode number */
+        off64_t        d_off;    /* 64-bit offset to next structure */
+        unsigned short d_reclen; /* Size of this dirent */
+        unsigned char  d_type;   /* File type */
+        char           d_name[]; /* Filename (null-terminated) */
+    };
+
+    long count = ::syscall(__NR_getdents64, fd, reinterpret_cast<linux_dirent64*>(buffer), buffer_size);
+    if (count == -1) {
+        if (errno == ENOENT) {
+            return ret;
+        }
+        throw std::system_error(std::error_code(errno, std::system_category()));
+    }
+
+    for (long next = 0; next < count; ) {
+        auto de = reinterpret_cast<linux_dirent64*>(buffer + next);
+        next += de->d_reclen;
+        std::optional<directory_entry_type> type;
+        switch (de->d_type) {
+        case DT_BLK:
+            type = directory_entry_type::block_device;
+            break;
+        case DT_CHR:
+            type = directory_entry_type::char_device;
+            break;
+        case DT_DIR:
+            type = directory_entry_type::directory;
+            break;
+        case DT_FIFO:
+            type = directory_entry_type::fifo;
+            break;
+        case DT_REG:
+            type = directory_entry_type::regular;
+            break;
+        case DT_LNK:
+            type = directory_entry_type::link;
+            break;
+        case DT_SOCK:
+            type = directory_entry_type::socket;
+            break;
+        default:
+            // unknown, ignore
+            ;
+        }
+        sstring name = de->d_name;
+        if (name == "." || name == "..") {
+            continue;
+        }
+        ret.emplace_back(directory_entry{std::move(name), type});
+    }
+
+    return ret;
+}
+
 future<> lister::rmdir(fs::path dir) {
```

Comment:
Patch title says it's `mkdir` re-implementation

Reply:
fixed